### ISSUE TITLE
Check current Cooperator status

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5937]
|<img alt="chris-'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| chris- reported|
|-|-|
|Trac ID|trac#5937|
|Type|enhancement|
|Created|2012-09-03 13:58:42Z|
|Branch|https://github.com/twisted/twisted/tree/cooperator-running-5937|

It is occasionally useful to know the current running status of a Cooperator. Everything is there except a public API. Attached patch remedies this.

Attachments:

* [cooperator_running.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b9/0b9df90e1cb845cfd7b53ce024fdaa47a500299b/7562b7fc1d9bf60ba5d3b718aa32a5f5805ee251.patch) (1496 bytes) - added by chris- on 2012-09-03 13:59:03Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5937 5937
type__enhancement enhancement
reporter__chris_ chris-
priority__normal normal
milestone__ 
branch__branches_cooperator_running_5937 branches/cooperator-running-5937
branch_author__chris___jml chris-, jml
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1346680722000000 1346680722000000
changetime__1361025693000000 1361025693000000
version__None None
owner__jml jml

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml set owner to @jml|
|-|-|

Thanks for the patch!

A few tweaks:

 1. The first line of the docstring for `test_running` is redundant.  Also, as a matter of personal taste, phrasing without "should" can make test docstrings read better.

 1. In the news patch, "determin" should be "determine".

I'm happy to make these changes and land the patch.  I won't be able to for another 5-6 hours though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

What about also adding some documentation for this feature?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

I think the docstring in the patch is enough for this feature.

I looked, but couldn't find any existing docs on `Cooperator` to update.  Adding such docs deserves its own ticket.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_5"></a>|@jml commented|
|-|-|

Filed [#6213](https://github.com/twisted/twisted/issues/6213) for docs for `Cooperator`.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

(In [37065]) Branching to 'cooperator-running-5937'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

(In [37067]) Docstring and news changes as per review (refs [#5937](https://github.com/twisted/twisted/issues/5937))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

(In [37068]) Split the 'running' test up into three tests.

Because that's how I roll (also, intermediate assertions aren't a
great thing)

(refs [#5937](https://github.com/twisted/twisted/issues/5937))

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_9"></a>|@jml removed owner|
|-|-|

Applied notes from my earlier review.  Once the tests were up in my editor I couldn't help splitting them up, so this needs another review.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince set owner to @jml|
|-|-|

  1. From reading the ticket, I assumed that `.running` would be an attribute. Also, there are already a number of places (`LoopingCall` which is also in `t.i.task` comes to mind) which have `.running` as attribute.
    * This involves turning it into a property.
  1. I think it would reasonable to have tests of `running` when the default value of `started=True` is passed.
Please resubmit for review after addressing the above points.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

(In [37145]) Make `running` a property, as per review item 1. refs [#5937](https://github.com/twisted/twisted/issues/5937).

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml set owner to @tomprince|
|-|-|

Thanks for the review!

1. Addressed in [37145] by making it a property.  Note the docstring change and the use of the rare (but now permitted?) decorator syntax.
2. Addressed in [37144].

I removed a pyflakes error from another test in the file while I was at it [37143].

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_13"></a>|@tomprince set owner to @jml|
|-|-|

  1. Please restore the `@return` annotation and add an `@rtype` annotation. Pydoctor currently documents properties as methods (see [generated docs](https://buildbot.twistedmatrix.com/builds/apidocs/apidocs-37145.tar.bz2)) and even when it supports properties, it should be able to deal with `@return`.
  1. I missed this on my first review, `assertTrue` and `assertFalse` should be preferred when appropriate.
Please merge after addressing the above points.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml commented|
|-|-|

(In [37151]) Use @return, because pydoctor prefers it. refs [#5937](https://github.com/twisted/twisted/issues/5937)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_15"></a>|@jml commented|
|-|-|

(In [37152]) Use assertTrue and assertFalse instead of assertEquals.  refs [#5937](https://github.com/twisted/twisted/issues/5937)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_16"></a>|@jml set owner to @tomprince|
|-|-|

Thanks for the quick turnaround. I've made the changes as you suggest.

Where does it say that `assertTrue(x)` is preferable to `assertEqual(x, True)`?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_17"></a>|@glyph commented|
|-|-|

Replying to [jml](#note_16):
> Where does it say that `assertTrue(x)` is preferable to `assertEqual(x, True)`?

If this _is_ official policy, we might want to just change the policy.  `assertEqual(x, True)` produces a nicer failure message.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_18"></a>|@jml commented|
|-|-|

Replying to [glyph](#note_17):
> Replying to [jml](#note_16):
> > Where does it say that `assertTrue(x)` is preferable to `assertEqual(x, True)`?
> 
> If this _is_ official policy, we might want to just change the policy.  `assertEqual(x, True)` produces a nicer failure message.
> 

I think it is too small a thing to be in a policy at all.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_19"></a>|@tomprince commented|
|-|-|

Or make `assertTrue` produce a better message? I've noticed that python 2.7's `TestCase` produces nicer message than trial's, at least for `assertTrue` and assertEqual`.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_20"></a>|@tomprince set owner to @jml|
|-|-|

Looks good, merge it.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_21"></a>|@jml set status to `closed`|
|-|-|

(In [37153]) Merge cooperator-running-5937: Add `running` property to `Cooperator`

Author: chris-, jml
Reviewers: jml, tom.prince
Fixes: [#5937](https://github.com/twisted/twisted/issues/5937)

`Cooperator` now has a boolean property, `running`, which can be used
to determine whether or not that cooperator is running.
